ensureInitializationComplete must be called after startInitialization

Hi. I am trying to get SMS details when the app is in background or not running. after I receive the SMS, the app crashes in bckground. Please help me fix this.

**STEPS RO REPRODUCE**
1. clone the Telephony Example project
2. run the project in VSCODE without debug mode.
3. Receive SMS while in app or while the app is in background (They give the same error).

I run this on android 8 on LG H990 (LG V20)

Then I get this error:

    E/AndroidRuntime(25457): FATAL EXCEPTION: main
    E/AndroidRuntime(25457): Process: com.example.sms, PID: 25457
    E/AndroidRuntime(25457): java.lang.RuntimeException: Unable to start receiver com.shounakmulay.telephony.sms.IncomingSmsReceiver: java.lang.IllegalStateException: ensureInitializationComplete must be called after startInitialization
    E/AndroidRuntime(25457): 	at android.app.ActivityThread.handleReceiver(ActivityThread.java:3329)
    E/AndroidRuntime(25457): 	at android.app.ActivityThread.-wrap17(Unknown Source:0)
    E/AndroidRuntime(25457): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1696)
    E/AndroidRuntime(25457): 	at android.os.Handler.dispatchMessage(Handler.java:105)
    E/AndroidRuntime(25457): 	at android.os.Looper.loop(Looper.java:164)
    E/AndroidRuntime(25457): 	at android.app.ActivityThread.main(ActivityThread.java:6710)
    E/AndroidRuntime(25457): 	at java.lang.reflect.Method.invoke(Native Method)
    E/AndroidRuntime(25457): 	at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
    E/AndroidRuntime(25457): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:770)
    E/AndroidRuntime(25457): Caused by: java.lang.IllegalStateException: ensureInitializationComplete must be called after startInitialization
    E/AndroidRuntime(25457): 	at io.flutter.embedding.engine.loader.FlutterLoader.ensureInitializationComplete(FlutterLoader.java:174)
    E/AndroidRuntime(25457): 	at com.shounakmulay.telephony.sms.IncomingSmsReceiver.processInBackground(IncomingSmsHandler.kt:77)
    E/AndroidRuntime(25457): 	at com.shounakmulay.telephony.sms.IncomingSmsReceiver.processIncomingSms(IncomingSmsHandler.kt:69)
    E/AndroidRuntime(25457): 	at com.shounakmulay.telephony.sms.IncomingSmsReceiver.onReceive(IncomingSmsHandler.kt:46)
    E/AndroidRuntime(25457): 	at android.app.ActivityThread.handleReceiver(ActivityThread.java:3318)
    E/AndroidRuntime(25457): 	... 8 more

Flutter Doctor output:

    Doctor summary (to see all details, run flutter doctor -v):
    [√] Flutter (Channel stable, 1.22.0, on Microsoft Windows [Version 10.0.18363.1198], locale en-US)
     
    [√] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    [√] Android Studio (version 3.6)
    [√] VS Code (version 1.51.1)
    [√] Connected device (1 available)

    • No issues found!

--- a/android/src/main/kotlin/com/shounakmulay/telephony/sms/IncomingSmsHandler.kt
+++ b/android/src/main/kotlin/com/shounakmulay/telephony/sms/IncomingSmsHandler.kt
@@ -22,6 +22,7 @@ import com.shounakmulay.telephony.utils.Constants.SHARED_PREFS_DISABLE_BACKGROUN
 import com.shounakmulay.telephony.utils.Constants.STATUS
 import com.shounakmulay.telephony.utils.Constants.TIMESTAMP
 import com.shounakmulay.telephony.utils.SmsAction
+import io.flutter.FlutterInjector
 import io.flutter.embedding.engine.FlutterEngine
 import io.flutter.embedding.engine.FlutterJNI
 import io.flutter.embedding.engine.dart.DartExecutor
@@ -73,9 +74,8 @@ class IncomingSmsReceiver : BroadcastReceiver() {
 
   private fun processInBackground(context: Context, sms: SmsMessage) {
     IncomingSmsHandler.apply {
-      backgroundContext = context
-      flutterLoader.ensureInitializationComplete(context.applicationContext, null)
       if (!isIsolateRunning.get()) {
+        initialize(context)
         val preferences = backgroundContext.getSharedPreferences(SHARED_PREFERENCES_NAME, Context.MODE_PRIVATE)
         val backgroundCallbackHandle = preferences.getLong(SHARED_PREFS_BACKGROUND_SETUP_HANDLE, 0)
         startBackgroundIsolate(backgroundContext, backgroundCallbackHandle)
@@ -113,12 +113,12 @@ fun SmsMessage.toMap(): HashMap<String, Any?> {
 object IncomingSmsHandler : MethodChannel.MethodCallHandler {
 
   internal val backgroundMessageQueue = Collections.synchronizedList(mutableListOf<SmsMessage>())
-  internal var flutterLoader = FlutterLoader.getInstance()
   internal var isIsolateRunning = AtomicBoolean(false)
 
   internal lateinit var backgroundContext: Context
   private lateinit var backgroundChannel: MethodChannel
   private lateinit var backgroundFlutterEngine: FlutterEngine
+  private lateinit var flutterLoader: FlutterLoader
 
   private var backgroundMessageHandle: Long? = null
 
@@ -129,7 +129,6 @@ object IncomingSmsHandler : MethodChannel.MethodCallHandler {
    * Also initializes the method channel on the android side
    */
   fun startBackgroundIsolate(context: Context, callbackHandle: Long) {
-    flutterLoader.ensureInitializationComplete(context, null)
     val appBundlePath = flutterLoader.findAppBundlePath()
     val flutterCallback = FlutterCallbackInformation.lookupCallbackInformation(callbackHandle)
 
@@ -148,7 +147,7 @@ object IncomingSmsHandler : MethodChannel.MethodCallHandler {
    * If any SMS were received during the background isolate was being initialized, it will process
    * all those messages.
    */
-  fun onInitialized() {
+  fun onChannelInitialized() {
     isIsolateRunning.set(true)
     synchronized(backgroundMessageQueue) {
 
@@ -178,6 +177,20 @@ object IncomingSmsHandler : MethodChannel.MethodCallHandler {
     args[HANDLE] = backgroundMessageHandle
     args[MESSAGE] = message
     backgroundChannel.invokeMethod(HANDLE_BACKGROUND_MESSAGE, args)
+  }
+
+  /**
+   * Gets an instance of FlutterLoader from the FlutterInjector, starts initialization and
+   * waits until initialization is complete.
+   *
+   * Should be called before invoking any other background methods.
+   */
+  internal fun initialize(context: Context) {
+    val flutterInjector = FlutterInjector.instance()
+    backgroundContext = context
+    flutterLoader = flutterInjector.flutterLoader()
+    flutterLoader.startInitialization(backgroundContext)
+    flutterLoader.ensureInitializationComplete(context.applicationContext, null)
   }
 
   fun setBackgroundMessageHandle(context: Context, handle: Long) {
@@ -224,7 +237,7 @@ object IncomingSmsHandler : MethodChannel.MethodCallHandler {
 
   override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
     if (SmsAction.fromMethod(call.method) == SmsAction.BACKGROUND_SERVICE_INITIALIZED) {
-      onInitialized()
+      onChannelInitialized()
     }
   }
 }

--- a/android/src/main/kotlin/com/shounakmulay/telephony/sms/SmsMethodCallHandler.kt
+++ b/android/src/main/kotlin/com/shounakmulay/telephony/sms/SmsMethodCallHandler.kt
@@ -203,7 +203,7 @@ class SmsMethodCallHandler(private val context: Context, private val smsControll
         IncomingSmsHandler.setBackgroundMessageHandle(context, backgroundHandle)
       }
       SmsAction.BACKGROUND_SERVICE_INITIALIZED -> {
-        IncomingSmsHandler.onInitialized()
+        IncomingSmsHandler.onChannelInitialized()
       }
       SmsAction.DISABLE_BACKGROUND_SERVICE -> {
         val preferences = context.getSharedPreferences(SHARED_PREFERENCES_NAME, Context.MODE_PRIVATE)

